Plugins/agspalrender/palrender.h

```diff
@@ -0,0 +1,194 @@
+#ifndef __PALGORITHMS_PALRENDER_H
+#define __PALGORITHMS_PALRENDER_H
+
+#include "agsplugin.h"
+#include <stdint.h> // for int32_t types, __int32 is just Windows
+#include <algorithm>  // for std::max and std::min
+#include <cstring>  // for memcpy
+
+
+#ifndef _WIN32
+#define __forceinline __attribute__((always_inline))
+
+#define SCRIPT_FLOAT(x) int32_t __script_float##x
+#define INIT_SCRIPT_FLOAT(x) float x; std::memcpy(&x, &__script_float##x, sizeof(float))
+#define FLOAT_RETURN_TYPE int32_t
+#define RETURN_FLOAT(x) int32_t __ret##x; std::memcpy(&__ret##x, &x, sizeof(float)); return __ret##x
+#else
+#define SCRIPT_FLOAT(x) __int32 __script_float##x
+#define INIT_SCRIPT_FLOAT(x) float x; memcpy(&x, &__script_float##x, sizeof(float))
+#define FLOAT_RETURN_TYPE __int32
+#define RETURN_FLOAT(x) __int32 __ret##x; memcpy(&__ret##x, &x, sizeof(float)); return __ret##x
+#endif

```

Comment:
Not related to this particular pull request, but cannot the use of union improve these memcpy hacks?
I recall we had a talk about this years ago. Certain systems (like mips) had issues when accessing variable of one type using a pointer to another, so JJS replaced these with memcpy.

E.g. original conversion from CJ's engine looked like:
`#define INIT_SCRIPT_FLOAT(x) float x = *((float*)&__script_float##x)`
then it became
`#define INIT_SCRIPT_FLOAT(x) float x; std::memcpy(&x, &__script_float##x, sizeof(float))`

We currently have a whole set of functions dedicated to systems with memory alignment problems:
https://github.com/adventuregamestudio/ags/blob/master/Common/util/memory.h#L57

Unfortunately my memory is vague about what is permitted to do on unaligned system and what is not. But considering our script interpreter works, there we have something like this:

<pre>
union Value
{
    int32_t iVal;
    float32_t fVal;
}
</pre>

And converting from int to float may work like (in a verbose way):
<pre>Value v;
v.iValue = iarg;
float f = v.fValue;
</pre>

Oh right, we have something similar in utility functions too (I copied this from SDL2):
https://github.com/adventuregamestudio/ags/blob/master/Common/util/bbop.h#L63

Reply:
afaict the current implementation is the only correct way to do it, because C requires a type-specific alignment for struct/union members and pointers to a type
http://port70.net/~nsz/c/c99/n1256.html#6.7.2.1p12
i don't know about C++, but usually they follow the C spec closely on the lower level

Reply:
I left the source mostly as is. I didn't want to stray too much from scavenger's work.

Reply:
No, I was only asking, because I have only recently similar code from the engine source.
I wonder where did they learn to use this code though. Was it copied from the engine or some other plugin?
Running search on our plugin sources, there's same code in lua plugin.